polytracker/src/taintdag/polytracker.cpp

```diff
@@ -7,212 +7,163 @@
  */
 
 #include "taintdag/polytracker.h"
-
+#include "taintdag/util.h"
 #include <sanitizer/dfsan_interface.h>
 
 #include <sys/stat.h>
 
-#include "taintdag/error.hpp"
+#include "taintdag/error.h"
 #include "taintdag/fnmapping.h"
 #include "taintdag/fntrace.h"
 
-namespace fs = std::filesystem;
-
 namespace taintdag {
 
-namespace details {
-// Returns the file size if it is usable to us (not error, > 0, <= max_label)
-std::optional<size_t> file_size(int fd) {
-  struct stat st;
-  if (fstat(fd, &st) == 0)
-    if (st.st_size > 0 && st.st_size <= max_label + 1)
-      return st.st_size;
-
-  return {};
-}
-
-bool reuse_prealloc_labels() {
-  auto v = getenv("POLYTRACKER_NO_REUSE_PREALLOC_LABELS");
-  return v == nullptr || v[0] != '0';
-}
-
-} // namespace details
-
 PolyTracker::PolyTracker(std::filesystem::path const &outputfile)
-    : of_{outputfile}, fdm_{of_.fd_mapping_begin(), of_.fd_mapping_end()},
-      fnm_{of_.fn_mapping_begin(), of_.fn_mapping_end()},
-      fnt_{of_.fn_trace_begin(), of_.fn_trace_end()},
-      tdag_{of_.tdag_mapping_begin(), of_.tdag_mapping_end()},
-      sinklog_{of_.sink_mapping_begin(), of_.sink_mapping_end()} {}
-
-PolyTracker::~PolyTracker() {
-  of_.fileheader_fd_count(fdm_.get_mapping_count());
-  of_.fileheader_tdag_size(tdag_.label_count() * sizeof(storage_t));
-  of_.fileheader_sink_size(sinklog_.size());
-  of_.fileheader_fn_count(fnm_.get_mapping_count());
-  of_.fileheader_trace_count(fnt_.get_event_count());
-}
+    : output_file_{outputfile} {}
 
 label_t PolyTracker::union_labels(label_t l1, label_t l2) {
-  auto ret = tdag_.union_taint(l1, l2);
-  return ret;
+  return output_file_.section<Labels>().union_taint(l1, l2);
 }
 
-void PolyTracker::open_file(int fd, fs::path const &path) {
-  // TODO (hbrodin): What if you open the same path twice? If we assume the file
-  // hasn't changed in between we should be able to reuse previously reserved
-  // source labels. Is this something we want to do? Or is it better to just
-  // generate new ranges on every open?
-  // TODO (hbrodin): This code is a a bit shaky. The main issue is with
-  // reserving labels that are not immediately initialized/assigned.
-  std::optional<taint_range_t> range;
-  bool reuse_source_range = details::reuse_prealloc_labels();
-  bool was_reused = false;
-  if (reuse_source_range) {
-    range = fdm_.existing_label_range(path.string());
-    was_reused = range.has_value();
-  }
-
-  if (!range) {
-    auto fsize = details::file_size(fd);
-    if (fsize) {
-      range = tdag_.reserve_source_labels(fsize.value());
+void PolyTracker::open_file(int fd, std::filesystem::path const &path) {
+  // Try to determine the file size. If unknown, just mark it as unknown.
+  uint64_t size = [fd]() -> uint64_t {
```

Comment:
Can we replace the initialization lambda with [std::filesystem::file_size](https://en.cppreference.com/w/cpp/filesystem/file_size) ?

Reply:
For sure! Should have been done that way in the first place...